### PR TITLE
Ruby: Add query for insecure mass assignment

### PR DESCRIPTION
**Pull request by joefarebrother**
This query looks for instances of request parameters having `permit!` called on them, which allows arbitrary parameters to be specified by the request, and then being used for a mass assignment operation.

### CODE REVIEW

File: ruby/ql/src/queries/security/cwe-915/MassAssignment.qhelp
Comment (hmac) [inline ruby/ql/src/queries/security/cwe-915/MassAssignment.qhelp:8]:
```suggestion
            allow arbitrary parameters to be set by the user. Otherwise, unintended attributes may be set, such as an <code>is_admin</code> field for a <code>User</code> object.
```

File: ruby/ql/src/queries/security/cwe-915/MassAssignment.qhelp
Comment (hmac) [inline ruby/ql/src/queries/security/cwe-915/MassAssignment.qhelp:31]: We could add a link to [this doc on Strong Parameters](https://guides.rubyonrails.org/action_controller_overview.html#strong-parameters)

**Review by hmac — COMMENTED** [review]
This looks great. I have only some minor comments. DCA shows a significant slowdown for one big Rails repo. I think it's worth running against the `rails-projects` source suite to see if there's a pattern there that we need to look into.

I also noticed in the Rails guide that you can do
```rb
params.permit(preferences: {})
```
which allows arbitrary data inside `preferences`. We should probably extend the query to cover that case, but that can be a follow-up and doesn't need to block this PR.

**Review by hmac — DISMISSED** [review]
👌

File: ruby/ql/lib/codeql/ruby/frameworks/ActiveRecord.qll
Comment (asgerf) [inline ruby/ql/lib/codeql/ruby/frameworks/ActiveRecord.qll:809]:
```suggestion
  private class MassAssignmentSink extends MassAssignment::Sink {
```

File: ruby/ql/lib/codeql/ruby/security/MassAssignmentQuery.qll
Comment (asgerf) [inline ruby/ql/lib/codeql/ruby/security/MassAssignmentQuery.qll:49]: Totally unsubstantiated guess, but perhaps performance would improve if you do the same thing for `isBarrierOut`? That is,
```codeql
predicate isBarrierOut(DataFlow::Node node, FlowState state) {
  isSink(node, state)
}
```

Calls to `update` in particular are sinks as well as having flow-through due to its summary in `Hash.qll`.

**Review by asgerf — COMMENTED** [review]
I came looking out of curiosity, looks great, but I'll defer to @hmac's review. Just a few drive-by comments

File: ruby/ql/lib/codeql/ruby/security/MassAssignmentCustomizations.qll
Comment (hvitved) [inline ruby/ql/lib/codeql/ruby/security/MassAssignmentCustomizations.qll:58]: It's better to use CFG nodes instead of AST nodes here. I.e., use `HashLiteralCfgNode` instead of `HashLiteral`, etc.

**Review by hvitved — APPROVED** [review]

**Pull request merged**
